### PR TITLE
SslStream override FlushAsync

### PR DESCRIPTION
**Pull request by benaadams**
SslStream should override FlushAsync over its internal stream; otherwise the default Stream implementation is used; which creates a new task which runs the sync/blocking flush and queues it to the threadpool.

Default Stream implementation:
```csharp
public virtual Task FlushAsync(CancellationToken cancellationToken)
{
	return Task.Factory.StartNew(state => ((Stream)state).Flush(), this,
		cancellationToken, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
}
```

Resolves https://github.com/dotnet/corefx/issues/14038

/cc @stephentoub

### CODE REVIEW

File: src/System.Net.Security/src/System/Net/Security/NegotiateStream.cs [hunk 1]
```diff
@@ -502,6 +502,21 @@ public override void Flush()
 #endif
         }
 
+#if DEBUG
+        public override async Task FlushAsync(CancellationToken cancellationToken)
+        {
+            using (DebugThreadTracking.SetThreadKind(ThreadKinds.User | ThreadKinds.Async))
+            {
+                await InnerStream.FlushAsync(cancellationToken);
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/NegotiateStream.cs]: .ConfigureAwait(false)

File: src/System.Net.Security/ref/System.Net.Security.cs [hunk 2]
```diff
@@ -160,6 +161,7 @@ public partial class SslStream : AuthenticatedStream
         public override int EndRead(IAsyncResult asyncResult) { throw null; }
         public override void EndWrite(IAsyncResult asyncResult) { }
         public override void Flush() { }
+        public override System.Threading.Tasks.Task FlushAsync(System.Threading.CancellationToken cancellationToken) { throw null; }
```
Comment (stephentoub) [inline src/System.Net.Security/ref/System.Net.Security.cs]: The src changes look good.

I think technically these ref changes are considered new APIs, even though they're just overrides, and need to be ifdef'd for netcoreapp11. @weshaggard?

**Comment by stephentoub** [comment]
@dotnet-bot test this please

File: src/System.Net.Security/src/System/Net/Security/NegotiateStream.cs (hunk 1 above)
Comment (benaadams) [inline src/System.Net.Security/src/System/Net/Security/NegotiateStream.cs]: Added

File: src/System.Net.Security/ref/System.Net.Security.cs [hunk 3]
```diff
@@ -160,6 +163,9 @@ public partial class SslStream : AuthenticatedStream
         public override int EndRead(IAsyncResult asyncResult) { throw null; }
         public override void EndWrite(IAsyncResult asyncResult) { }
         public override void Flush() { }
+#if netcoreapp11
+        public override System.Threading.Tasks.Task FlushAsync(System.Threading.CancellationToken cancellationToken) { throw null; }
+#endif
```
Comment (stephentoub) [inline src/System.Net.Security/ref/System.Net.Security.cs]: Thanks, @benaadams. Do we have tests covering these already?

**Comment by Priya91** [comment]
cc @weshaggard Does implementing override require api-review?

**Comment by weshaggard** [comment]
>@weshaggard Does implementing override require api-review?

No. Strictly speaking we could probably get away without #ifdef'ing them in the ref either but I'm OK either way.

**Comment by stephentoub** [comment]
> Since it's already deriving from Stream the ref folder should either not be touched or the APIs should be added in netstandard as well (no #if netcoreapp11

If we don't include it in the ref, then it's possible some call sites could skip the derived implementation with a direct call to the base.  If we're ok with that because they're corner case (e.g. a type derived from SslStream calling base.FlushAsync), ok.

> Strictly speaking we could probably get away without #ifdef'ing them in the ref either but I'm OK either way.

Wouldn't it then be possible for the C# compiler to emit some call sites to the method based on them being in the ref and then find they're not overridden in some implementations?

**Comment by weshaggard** [comment]
>Wouldn't it then be possible for the C# compiler to emit some call sites to the method based on them being in the ref and then find they're not overridden in some implementations?

I've not tested this but I believe that would still work. I'm 100% sure we have overrides in some of our refs that only exist in one of our platforms and not all.

**Comment by stephentoub** [comment]
Ok, thanks, Wes.

**Comment by benaadams** [comment]
So should work without the ref changes; the ref changes will ensure derived classes also work in 1.1+?

Still required
* Need to check tests

That correct?

**Comment by stephentoub** [comment]
> That correct?

Yup.  If you could add a test for each, that'd be great.  I took a quick look at a code coverage report with your change and they don't appear to have any.

**Comment by stephentoub** [comment]
@benaadams, are you still working on this?

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows_NT Debug please
@dotnet-bot test Outerloop Ubuntu14.04 Debug please

File: src/System.Net.Security/src/System/Net/Security/NegotiateStream.cs [hunk 4]
```diff
@@ -502,6 +502,11 @@ public override void Flush()
 #endif
         }
 
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            return InnerStream.FlushAsync(cancellationToken);
```
Comment (benaadams) [inline src/System.Net.Security/src/System/Net/Security/NegotiateStream.cs]: @stephentoub the Async test fails when I add the 
```
using (DebugThreadTracking.SetThreadKind(ThreadKinds.User | ThreadKinds.Async))
{
    await InnerStream.FlushAsync(cancellationToken);
}
```
And the task doesn't complete

Are they the wrong flags?

Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/NegotiateStream.cs]: I've not used that, but my guess is that it was never intended to be used across threads.

Comment (benaadams) [inline src/System.Net.Security/src/System/Net/Security/NegotiateStream.cs]: Ah, different type of async. K, should work as now.

**Comment by benaadams** [comment]
Also added tests for sync/blocking Flush as didn't seem to be any either

**Comment by benaadams** [comment]
@dotnet-bot test Outerloop Windows_NT Debug please
@dotnet-bot test Outerloop Ubuntu14.04 Debug please

**Comment by benaadams** [comment]
@dotnet-bot test Outerloop Windows_NT Debug please

**Comment by benaadams** [comment]
New tests pass and exercise Flush and FlushAsync

**Pull request merged**